src/cudamatrix/cu-matrix.h

```diff
@@ -693,14 +694,42 @@ class CuMatrixBase {
 
   // The following two functions should only be called if we did not compile
   // with CUDA or could not get a CUDA card; in that case the contents are
-  // interpreted the same as a regular matrix.  DON'T USE THESE UNLESS YOU KNOW
-  // WHAT YOU ARE DOING!
+  // interpreted the same as a regular matrix.  Don't use these unless you know
+  // what you are doing!
   inline const MatrixBase<Real> &Mat() const {
     return *(reinterpret_cast<const MatrixBase<Real>* >(this));
   }
   inline MatrixBase<Real> &Mat() {
     return *(reinterpret_cast<MatrixBase<Real>* >(this));
   }
+  
+  //////////////////////////////////////////////////////
+  ////           FSMN kernel functions          ///////
+  ////////////////////////////////////////////////////
+
+  // forward operation in memory block
```

Comment:
If we were to commit this stuff I'd probably want these to be non-member functions as they are pretty specialized-- maybe in cu-math.h.    But see my overall review.

Reply:
Thanks a lot for the review and suggestion.  I am agree that add these DFSMN specialized kernel function is conflict with the code style of kaldi. We will try to think of some alternative methods although it may not efficient as the CUDA kernel function.

Reply:
I was just saying they should be non-member functions, they would still be CUDA kernels.
But that was a minor point, and fixing it would likely not make me want to merge this.
My problem is a little more fundamental-- it's that there is a lot of code, it's based on a setup that is basically deprecated, and it provides no clear advantage versus our current best models based on TDNN-F.